### DISCUSSION TITLE
IAsyncLifetime and IClassFixture with Testcontainers

### DISCUSSION THREAD

**Discussion by jhensley707** [xunit/xunit#2834] [Question] [answered]
I read the Testcontainers documentation for .NET for implementing a PostgreSql database container in Xunit tests. The suggestion was to use IAsyncLifetime to spin up a docker container for each test. The first few tests worked very well. As I increased the number of tests, I suddenly started getting timeout issues trying to start the test container. So that might be an issue with Testcontainers. Or perhaps just my implementation.

If I could not implement this as an Xunit shared context in a single test class, I thought I might try a Class Fixture, which should allow me to create one instance to manage the container for all the tests. Unfortunately, the Testcontainer documentation starts the docker container in the InitializeAsync() method, which gets run for every test. Thus, refactoring my initial design for a class fixture offered no change in the behavior. About 20 of my 75 tests time out.

Starting the Testcontainer is an async method which is why it works well in the InitializeAsync method. But this does not seem scalable for many tests to cover all my repository methods.

Is there a way I could get Xunit tests to asynchronously start the test container once and use it for all the tests? I can't imagine how to call the StartAsync method in the Class Fixture constructor so that the Testcontainer is only created one time for the tests, then cleaned up at the end of all the tests.

It is almost seems like I need an InitializeFixtureAsync method for fixture creation and then the InitializeAsync method for preparation of each test.

**Comment by bradwilson** [comment]
What you're describe that you want, is how collection fixtures should work. Can you provide a repro project?

**Reply by jhensley707** [reply]
My company would probably frown if I posted my current working prototype with business-specific entities but I will spend some time this weekend roughing up a relevant repro project.

**Reply by jhensley707** [reply]
I have created a repository with both shared context methods, one using Class Fixture and one using a common base class.

[Repro](https://github.com/jhensley707/XunitTestcontainerRepro)

The limited tests I created all work; I cannot get this to generate the same timeout error. However, the error occurs at
 `public async Task InitializeAsync() { await _postgres.StartAsync(); }`

I put a break point at each occurrence and run the tests in Debug. That line is hit in the Class Fixture instance for every test, rather than starting once and letting me work with the same database container through all tests.

**Reply by bradwilson** [reply, marked as answer]
Ah, I see. What you want is a collection fixture, not a class fixture, because you want a single shared instance of `DataContextFixture` across all the tests across multiple classes. https://xunit.net/docs/shared-context#collection-fixture

**Reply by jhensley707** [reply]
Thank you Brad,

I tried that implementation. I updated my [Repro ](https://github.com/jhensley707/XunitTestcontainerRepro) project with a new Collection Fixture test suite. When I put a breakpoint on the
`await _postgres.StartAsync();`
of my new DataContextCollectionFixture.InitializeAsync() method, I find once again that the method is hit for every test, rather than once at the beginning of all tests.
Have I done something wrong in that implementation?

**Reply by jhensley707** [reply]
Interestingly, I found that the IAsyncLifetime.InitializeAsync() does work on my Class Fixture as we discussed if I have a [Theory] with [InlineData()] running the test multiple times. But I only have a handful of those. The rest of my tests are [Fact].

**Reply by bradwilson** [reply]
Here is a repro to show how it's used:

```csharp
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Empty;

public class Fixture : IAsyncLifetime
{
	public static int DisposeCount;
	public static int InitializeCount;

	public Task DisposeAsync()
	{
		Interlocked.Increment(ref DisposeCount);
		return Task.CompletedTask;
	}

	public Task InitializeAsync()
	{
		Interlocked.Increment(ref InitializeCount);
		return Task.CompletedTask;
	}
}

[CollectionDefinition(nameof(MyCollection))]
public class MyCollection : ICollectionFixture<Fixture>
{ }

[Collection(nameof(MyCollection))]
public class TestClass1
{
	readonly Fixture fixture;

	public TestClass1(Fixture fixture) =>
		this.fixture = fixture;

	[Fact]
	public void TestMethod1()
	{
		Assert.Equal(1, Fixture.InitializeCount);
		Assert.Equal(0, Fixture.DisposeCount);
	}

	[Fact]
	public void TestMethod2()
	{
		Assert.Equal(1, Fixture.InitializeCount);
		Assert.Equal(0, Fixture.DisposeCount);
	}
}

[Collection(nameof(MyCollection))]
public class TestClass2
{
	readonly Fixture fixture;

	public TestClass2(Fixture fixture) =>
		this.fixture = fixture;

	[Fact]
	public void TestMethod1()
	{
		Assert.Equal(1, Fixture.InitializeCount);
		Assert.Equal(0, Fixture.DisposeCount);
	}

	[Fact]
	public void TestMethod2()
	{
		Assert.Equal(1, Fixture.InitializeCount);
		Assert.Equal(0, Fixture.DisposeCount);
	}
}
```

And here is the result of running the tests:

![image](https://github.com/xunit/xunit/assets/16944/6c456ffa-d5d9-4f34-bc59-73f5b4334ad1)

Where is the deviation from what you're doing?

**Reply by jhensley707** [reply]
In my Repro, the Collection definition is a class defined inside my Fixture. Your sample declares it separately. I think that might be important. The other thing I did not do was remove IClassFixture in my test base class definition. And like magic, the CollectionFixture works as you had indicated. Apologies!
